drivers/dma/dma_nxp_imx_edma.h

```diff
@@ -0,0 +1,478 @@
+ *	HALFMAJOR is enabled, the whole buffer otherwise).
+ *
+ *	- for producer channels, each DMA interrupt will
+ *	signal that an amount of data has been added
+ *	to the buffer.
+ *
+ *	- to signal this, the ISR uses EDMA_CHAN_PRODUCE_CONSUME_A
+ *	which will "consume" data from the buffer for
+ *	consumer channels and "produce" data for
+ *	producer channels.
+ *
+ *	- since the upper layers using this driver need
+ *	to let the EDMA driver know whenever they've produced
+ *	(in the case of consumer channels) or consumed
+ *	data (in the case of producer channels) they can
+ *	do so through the reload() function.
+ *
+ *	- reload() uses EDMA_CHAN_PRODUCE_CONSUME_B which
+ *	for consumer channels will "produce" data and
+ *	"consume" data for producer channels, thus letting
+ *	the driver know what action the upper layer has
+ *	performed (if the channel is a consumer it's only
+ *	natural that the upper layer will write/produce more
+ *	data to the buffer. The same rationale applies to
+ *	producer channels).
+ *
+ *	- EDMA_CHAN_PRODUCE_CONSUME_B is just the opposite
+ *	of EDMA_CHAN_PRODUCE_CONSUME_A. If one produces
+ *	data, the other will consume and vice-versa.
+ *
+ *	- all of this information is valid only in the
+ *	context of cyclic buffers. If this behaviour is
+ *	not enabled, querying the status will simply
+ *	resolve to querying CITER and BITER.
+ */
+#define EDMA_CHAN_PRODUCE_CONSUME_A(chan, size)\
+	((chan)->type == CHAN_TYPE_CONSUMER ?\
+	 edma_chan_cyclic_consume(chan, size) :\
+	 edma_chan_cyclic_produce(chan, size))
+
+#define EDMA_CHAN_PRODUCE_CONSUME_B(chan, size)\
+	((chan)->type == CHAN_TYPE_CONSUMER ?\
+	 edma_chan_cyclic_produce(chan, size) :\
+	 edma_chan_cyclic_consume(chan, size))
+
+enum channel_type {
+	CHAN_TYPE_CONSUMER = 0,
+	CHAN_TYPE_PRODUCER,
+};
+
+enum channel_state {
+	CHAN_STATE_INIT = 0,
+	CHAN_STATE_CONFIGURED,
+	CHAN_STATE_STARTED,
+	CHAN_STATE_STOPPED,
+	CHAN_STATE_SUSPENDED,
+};
+
+struct edma_channel {
+	uint32_t id;
+	const struct device *dev;
+	enum channel_state state;
+	enum channel_type type;
+	void *arg;
+	dma_callback_t cb;
+	int irq;
+	struct dma_status stat;
+	uint32_t bsize;
+	bool cyclic_buffer;
+};
+
+struct edma_data {
+	/* this needs to be the first member */
+	struct dma_context ctx;
+	mm_reg_t regmap;
+	struct edma_channel *channels;
+	atomic_t channel_flags;
+	struct edma_hal_config *hal_cfg;
+};
+
+struct edma_config {
+	uint32_t regmap_phys;
+	uint32_t regmap_size;
+	void (*irq_config)(void);
+	bool contiguous_channels;
+};
+
+static inline int channel_change_state(struct edma_channel *chan,
+				       enum channel_state next)
+{
+	enum channel_state prev = chan->state;
+
+	LOG_DBG("attempting to change state from %d to %d for channel %d", prev, next, chan->id);
+
+	/* validate transition */
+	switch (prev) {
+	case CHAN_STATE_INIT:
+		if (next != CHAN_STATE_CONFIGURED) {
+			return -EPERM;
+		}
+		break;
+	case CHAN_STATE_CONFIGURED:
+		if (next != CHAN_STATE_CONFIGURED &&
```

Comment:
I guess logically speaking there's nothing wrong with re-configuring an already configured channel. Not sure why you'd do this though.

For now, we're going to forbid this transition (assuming SOF doesn't call config() twice as it does for the DAI, which I'll have to test) because `edma_config()` will fail if called a second time while the channel state is != `STOPPED`. This is because the function setting the MUX value queries the current MUX value and if != 0 it will fail. The only function clearing the MUX value is `edma_stop()`, thus the need for the channel to be in `STOPPED` state for the second `edma_config()` call.

If we ever find a case in which `edma_config()` is called twice we can allow the transition and fix the MUX setting issue pretty easily.